### PR TITLE
Add `doc/user-guide/glossary` page

### LINKED ISSUE #1395: glossary: create Glossary page (besides Basic Concepts?)
"Basic Concepts" is an actual proposed section (see #550) and not related (?) specifically to the tooltips/glossary. Furthermore, it's used all over docs, not just in the User Guide.

Should we put it perhaps in content/docs/glossary/ ?

---

UPDATE: Jump to https://github.com/iterative/dvc.org/issues/1395#issuecomment-857923802

### PR DESCRIPTION
**Pull request by julieg18**
* creates a template called `doc-jsx.jsx` for creating documentation pages with jsx instead of markdown
* creates a `doc/user-guide/glossary` page that lists all tooltip data in this format:
```markdown
# Glossary

<b id="bold-term">Bold term</b>: description text
```
* see finished page at https://dvc-org-add-glossary-pa-jxqnju.herokuapp.com/doc/user-guide/glossary

Fixes #1395

@jorgeorpinel @casperdcl @iesahin

### CODE REVIEW

File: content/docs/sidebar.json
Comment (julieg18) [inline content/docs/sidebar.json:108]: I noticed that this page and the page on line `143` had incorrect `source` causing the "Edit on Github" link to take you to a 404 page. I went ahead and fixed them.

File: content/docs/sidebar.json
Comment (julieg18) [inline content/docs/sidebar.json:100]: I wasn't sure where to have the "Edit on Github" link take you since there isn't a single page. I linked the page to the [`basic-concepts`](https://github.com/iterative/dvc.org/tree/master/content/docs/user-guide/basic-concepts) folder for now.

Comment (casperdcl) [reply content/docs/sidebar.json:100]: makes most sense to me, yes.

File: src/pages/doc/user-guide/glossary.tsx
Comment (casperdcl) [inline src/pages/doc/user-guide/glossary.tsx:12]: ideally this would be case-insensitive sorted alphabetically by name

File: content/docs/sidebar.json
Comment (casperdcl) [inline content/docs/sidebar.json:101]: needs to be moved further down the sidebar, question is where? CC @iterative/docs

File: src/pages/doc/user-guide/glossary.tsx
Comment (julieg18) [reply src/pages/doc/user-guide/glossary.tsx]: Terms are now sorted!

File: src/pages/doc/user-guide/glossary.tsx
Comment (casperdcl) [inline src/pages/doc/user-guide/glossary.tsx:16]:
```suggestion
          return c1Name < c2Name ? -1 : c2Name < c1Name ? 1 : 0
```

Comment (julieg18) [reply src/pages/doc/user-guide/glossary.tsx]: Done!

**Review by rogermparent — COMMENTED** [review]
Looks good on the code end! I'll just give this a comment for now and let the docs team approve when the sidebar and any other requests are handled.

Nice work!

File: content/docs/sidebar.json
Comment (shcheklein) [reply content/docs/sidebar.json:101]: Agreed, that we should move it down. Probably even to the bottom of the UG?

**Comment by shcheklein** [comment]
One thing - let's try to generate 3rd or 2nd level titles? This way items will have links? @iterative/docs WDYT?

**Review by shcheklein — APPROVED** [review]

**Comment by iesahin** [comment]
Nice work @julieg18 Thank you. 

If I open the glossary link directly, it's like this: 
Screen Shot 2021-07-01 at 11 33 10

Link is: https://dvc-org-add-glossary-pa-jxqnju.herokuapp.com/doc/user-guide/glossary

if I navigate the page from the sidebar, it's like this: 
Screen Shot 2021-07-01 at 11 33 32

This is with Firefox 89 on macOS 11.4

In the first version, there is a sudden collapse after loading the page.

File: content/docs/sidebar.json
Comment (iesahin) [reply content/docs/sidebar.json:101]: I think Glossary can be the last section.

**Comment by casperdcl** [comment]
> let's try to generate 3rd or 2nd level titles? This way items will have links?

All entries already have `#id`s but I guess you mean there should be an on-hover link icon appear on the left of entries

**Comment by julieg18** [comment]
> This is with Firefox 89 on macOS 11.4
> 
> In the first version, there is a sudden collapse after loading the page.

Hmmmmm, I'm on macOS 11.2.3 and firefox 89.0.2 but I when I click on the link, the sidebar is showing. I tried safari and chrome as well. 

image

I also tried on my windows laptop, but still can't seem to replicate the bug. @iesahin, are you able to see the bug on a different browser? Also are the other documentation links broken as well or just the glossary page?

**Comment by iesahin** [comment]
That's not the sidebar, the content after **File Linking** is not shown if I click the link.

**Comment by julieg18** [comment]
> That's not the sidebar, the content after **File Linking** is not shown if I click the link.

Oh, I see. Thanks for the clarification!

File: content/docs/sidebar.json
Comment (julieg18) [reply content/docs/sidebar.json]: I moved glossary to the bottom of user-guide for now.

**Comment by julieg18** [comment]
Hyperlink is now added to heading(and can be added to any other headings on the page if we decide to do so) and all the glossary elements should be rendering properly!

**Comment by casperdcl** [comment]
> and can be added to any other headings on the page if we decide to do so

yes please (as in hovering on a term will show a link icon :link: on the left of it)

**Comment by julieg18** [comment]
> yes please (as in hovering on a term will show a link icon :link: on the left of it)

Do we want to convert the terms into h2s or mimic a headers link behavior while keeping the terms how they look now?

**Comment by casperdcl** [comment]
The "look" should be the same as now. Whether internally they are somehow h2s without line breaks or (probably easier) mimic link behaviour is up to you.

**Comment by julieg18** [comment]
The terms' names are now hyperlinks, acting like Header hyperlinks!

File: src/components/Documentation/index.tsx
Comment (casperdcl) [inline src/components/Documentation/index.tsx:28]: not too sure about purpose of this change but whatever...

**Review by casperdcl — APPROVED** [review]
awesome

File: src/components/Documentation/index.tsx
Comment (julieg18) [reply src/components/Documentation/index.tsx:28]: We need to use the GitHub link in two places now. The component that renders markdown and the component that renders JSX. To avoid using the link twice I just created a function that holds the link  and both components use the function. I can take it off if you like 😀

Comment (casperdcl) [reply src/components/Documentation/index.tsx:28]: Afaik you're only using the function in one place in this file... oh you mean you're using it elsewhere from this file?

Comment (casperdcl) [reply src/components/Documentation/index.tsx:28]: Ah I see yes you are...

File: content/docs/sidebar.json
Comment (jorgeorpinel) [reply content/docs/sidebar.json:103]: Interesting. Should the engine throw an error when building? Out of scope for this PR but could open an issue (or move the fix to another PR and we can decide there).

**Comment by jorgeorpinel** [comment]
I never noticed the Codeclimate check [failing](https://codeclimate.com/github/iterative/dvc.org/pull/2595) in docs repo PRs. Is it an issue?

File: content/docs/sidebar.json
Comment (jorgeorpinel) [inline content/docs/sidebar.json:181]: What would it take to reduce this to just `"glossary"`? Moving the `basic-concepts/` folder?

**Review by jorgeorpinel — COMMENTED** [review]
https://dvc-org-add-glossary-pa-jxqnju.herokuapp.com/doc/user-guide/glossary looks is great, thanks @julieg18! I'm just unsure where should it go in the nav but that's easy to change later.

I'm slightly worried that we won't have a good index for the actual Basic Concepts section (#550) per https://github.com/iterative/dvc.org/issues/1395#issuecomment-866648657. Are we sure we're not going to want to move this into there? Because that would require a permanent redirect at that point. Cc @casperdcl @shcheklein @iesahin 

Other than that, I'm curious about some code changes but I'll leave that for @rogermparent perhaps since I'm unable to determine exactly how they related to the Glossary page. Specifically:
- `src/components/Documentation/Markdown/inde/Main/` folder was created with 2 pre-existing files but unfortunately GH didn't pick up the move, so it's very hard to tell what changed in them.
- What is src/components/Documentation/WithJSX/AutoLinkElement/index.tsx ? (Line 67 looks fun)
- Same for src/templates/doc-jsx.tsx . Where is `IJSXDocPageProps` used?

Thanks

File: content/docs/sidebar.json
Comment (julieg18) [reply content/docs/sidebar.json:181]: I'm not sure if I'm understanding your question 🤔 Are you referring to deleting `source` or renaming the `source` value to `glossary`? The `source` is used to build the "Edit on Github" button. If you're asking about deleting the source, we don't have a `/content/docs/user-guide/glossary.md` so deleting the source would end up creating a 404 link. I linked the `basic-concepts` folder, since that is where we're currently holding all of our tooltip data. If you're asking about renaming the `source` to `glossary`, I suppose we could do that by renaming the `basic-concepts` folder to `glossary` 😃

**Comment by julieg18** [comment]
> * `src/components/Documentation/Markdown/inde/Main/` folder was created with 2 pre-existing files but unfortunately GH didn't pick up the move, so it's hard to tell what changed.

@jorgeorpinel `Markdown/index.tsx` originally took in `htmlAst` as props and rendered most of the documentation page. We needed to be able render that part of the documentation page with `JSX` or `htmlAst` so I took the majority of `Markdown/index.tsx` and placed it in a separate component called `Markdown/Main`. The `Main` component is used to render part of documentation page whether you want to render [JSX](https://github.com/iterative/dvc.org/pull/2595/files#diff-70b929b52d33ce75864c16088d554ac6fcdf09326f472ae6aa8a906ad30e3c92R24) or [htmlAST](https://github.com/iterative/dvc.org/pull/2595/files#diff-ff8e86061a1e884184e814bc602693222d3ab20a89c6ac157af3590bc4f846f3R127).

> * What is src/components/Documentation/WithJSX/AutoLinkElement/ ? (index.tsx:67 looks fun)

We have a `gatsby-plugin` that converted our markdown headings into hyper-links. To render hyper-links with JSX, I created the `AutoLinkElement`. It already has access to the needed css, so all I needed to do was create the html and add the right classes. 

> * Same for src/templates/doc-jsx.tsx . Where is IJSXDocPageProps used?

Whoops, that's a typo. I'll delete the `export`. Thanks for spotting that!

File: content/docs/sidebar.json
Comment (julieg18) [reply content/docs/sidebar.json:103]: I'm not sure about the engine throwing an error, but I don't mind moving these fixes to a separate PR if needed!

File: content/docs/sidebar.json
Comment (casperdcl) [reply content/docs/sidebar.json:181]: > What would it take to reduce this sidebar entry to just `"glossary"`? Moving the `basic-concepts/` folder?

I think by "sidebar entry" you meant page URL slug?

The sidebar entry *is* Glossary, and the page URL slug *is* "glossary." The "source" folder is only used for the "Edit on GH" button.

**Comment by casperdcl** [comment]
> Are we sure we're not going to want to move this into there? Because that would require a permanent redirect at that point

I don't understand. Is this related to [this confusion](https://github.com/iterative/dvc.org/pull/2595#discussion_r663195419)?

File: content/docs/sidebar.json
Comment (jorgeorpinel) [reply content/docs/sidebar.json:181]: I was referring to this:
```suggestion
      "glossary",
```

Comment (jorgeorpinel) [reply content/docs/sidebar.json:181]: > If you're asking about renaming the source to glossary, I suppose we could do that by renaming the basic-concepts folder to glossary 

Right. Should we do that? But again... #550

**Comment by jorgeorpinel** [comment]
Thanks for the explanations @julieg18! Seems like the code has been peer reviewed so 👌 

> > Are we sure we're not going to want to move this into there? Because that would require a permanent redirect at that point
> 
> I don't understand. Is this related to this confusion?

It's related to the naming somewhat, yes. But there's a reason we have glossary tests in `basic-concepts/` right? They're closely related. So what will we have as index for Basic Concetps? I know it's beyond the scope here technically but not addressing it now may mean having to do a redirect later, which we can do (I'm not trying to block this merge) but its not ideal.

I'll mention it in tomorrow's meeting.

File: content/docs/sidebar.json
Comment (julieg18) [reply content/docs/sidebar.json:181]: I don't think we'd be able to do that since then the "Edit on Github" link would default to https://github.com/iterative/dvc.org/blob/master/content/docs/user-guide/glossary.md which doesn't exist.

**Comment by julieg18** [comment]
Are we good to merge this, or is there is more to discuss?

**Comment by casperdcl** [comment]
I still don't know what @jorgeorpinel means by "redirect" (i.e. redirect what to what?) but I suspect by "redirect" he is referring to people bookmarking the "edit on GH" URL rather than any dvc.org URL.

1. Not something anyone does
2. Not something we need to care about
3. The source folder layouts are absolutely irrelevant, the only thing which matters is dvc.org URLs never changing or redirecting.

I also don't know what @jorgeorpinel means by "index" but I suspect that once again he means where in source code does a different issue (#550) get addressed, and the answer as repeatedly stated in both issues & meetings is in the same folder. The yml/md source files each will contain 2 keys (short & long descriptions: the short ones are rendered to Glossary, the long ones to BC #550).

Definite immediate merge to me unless @jorgeorpinel is envisioning a different source layout.

**Comment by jorgeorpinel** [comment]
UPDATE: OK, we discussed this in the meeting and we agreed to keep the glossary page where it is since it may take a while to get to #550 anyway. Also we can use it initially as the "Basic Concepts index" if we only have a few BCs (e.g. #2453) by linking them from this glossary page.

**Comment by jorgeorpinel** [comment]
Thanks again @ju0gri ! 🎉

File: content/docs/sidebar.json
Comment (jorgeorpinel) [reply content/docs/sidebar.json:181]: I just noticed this. Hmm so it's a special/hardcoded nav entry then. Not sure I love that but oh well, this is merged!

Comment (jorgeorpinel) [reply content/docs/sidebar.json:181]: Wait no I see. I think it could default to /content/docs/user-guide/glossary/index.md too? Again, this is merged so np but I was thinking what does it take to decouple it completely from `basic-concepts/`

**Comment by casperdcl** [comment]
> By redirect I mean [https://github.com/iterative/dvc.org/blob/master/redirects-list.json](https://github.com/iterative/dvc.org/blob/master/redirects-list.json?rgh-link-date=2021-07-07T16%3A00%3A09Z) i.e. an actual HTTP 301/303 redirect.

But neither `glossary` nor `basic-concepts` currently and in future will ever appear in that file?

> > dvc.org URLs never changing or redirecting
> 
> Exactly. I'm trying to prevent it from changing and redirecting if we have to move the `glossary/` to `basic-concepts/` (index) soon after this is merged.

completely lost

> By index I mean `index.md`, which is shown for parent nav entries e.g. [dvc.org/doc/start](https://dvc.org/doc/start) (content/doc/start/index.md)

https://github.com/iterative/dvc.org/pull/2595#discussion_r667363616:

> Wait no I see. I think it could default to /content/docs/user-guide/glossary/index.md too? Again, this is merged so np but I was thinking what does it take to decouple it completely from `basic-concepts/`

- There is no such thing as `/content/docs/user-guide/glossary/index.md`.
- There is no such thing as `/content/docs/user-guide/glossary/`.
- There is no such thing as `/content/docs/user-guide/basic-concepts/index.md`.

And none of the above 3 things will ever exist.

File: content/docs/sidebar.json
Comment (casperdcl) [reply content/docs/sidebar.json:181]: confusion related to https://github.com/iterative/dvc.org/pull/2595#issuecomment-877775897

**Pull request merged**
